### PR TITLE
add support for open-interpreter.

### PR DESCRIPTION
**Pull request by orange-crow**
添加对open-interpreter的支持，有两种方式使用open-interpreter：
Add support for open-interpreter, there are two ways to use open-interpreter:

1. 直接使用open-interpreter的原生能力：
 Directly use the native capabilities of open-interpreter:

```python
from metagpt.tools.code_interpreter import OpenCodeInterpreter
figure_path = './tests/data/figure_ci.png'
ci_ploter = OpenCodeInterpreter()
ci_ploter.chat(f"使用seaborn对{df_path}中与股票布林带有关的数据列的Date, Close, SMA, BB_upper（布林带上界）, BB_lower（布林带下界）进行可视化, 可视化图片保存在{figure_path}中。不需要任何指标计算，把Date列转换为日期类型。要求图片优美，BB_upper, BB_lower之间使用合适的颜色填充。")
```
2. 通过装饰器减少MG中action的代码工作量
Reduce the amount of code in action in MG through decorators

- 支持将open-interpreter各个plan中的代码片段拼接成函数，可以自动去除open-interpreter中的错误代码。
- 将生成的函数保存为脚本文件。
- 支持从脚本文件执行需要的函数。
```python
from metagpt.tools.code_interpreter import OpenInterpreterDecorator
from metagpt.actions import Action

class CreateStockIndicators(Action):
    @OpenInterpreterDecorator(save_code=True, code_file_path="./tests/data/stock_indicators.py")
    async def run(self, stock_path: str, indicators=['Simple Moving Average', 'BollingerBands']) -> pd.DataFrame:
        """对stock_path中的股票数据, 使用pandas和ta计算indicators中的技术指标, 返回带有技术指标的股票数据，不需要去除空值, 不需要安装任何包；
            指标生成对应的三列: SMA, BB_upper, BB_lower
        """
        ...
 
# test this class
indicators = ['Simple Moving Average', 'BollingerBands']
stocker = CreateStockIndicators()
df, msg = await stocker.run(stock, indicators=indicators)
# save df a csv file，we need the csv file path in next action.
df_path = './tests/data/stock_indicators.csv'
df.to_csv(df_path)
```

The content of ./tests/data/stock_indicators.py is as follows：
```python
# ./tests/data/stock_indicators.py
import pandas as pd
import ta

def run(stock_path: str, indicators=['Simple Moving Average', 'BollingerBands']) -> pd.core.frame.DataFrame:
    stock_data = pd.read_csv(stock_path)
    stock_data.head()

    if 'Simple Moving Average' in indicators:
        # Calculate Simple Moving Average (SMA)
        stock_data['SMA'] = ta.trend.sma_indicator(stock_data['Close'], window=20)

    if 'BollingerBands' in indicators:
        # Calculate Bollinger Bands
        bb_indicator = ta.volatility.BollingerBands(stock_data['Close'], window=20)
        stock_data['BB_upper'] = bb_indicator.bollinger_hband()
        stock_data['BB_lower'] = bb_indicator.bollinger_lband()

    stock_data.head()
    return stock_data
```

### CODE REVIEW

File: metagpt/actions/clone_function.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        code_path.parent.mkdir(parents=True, exist_ok=True)
+        code_path.write_text(code)
+        logger.info(f"Saving Code to {code_path}")
+
+    async def run(self, template_func: str, source_code: str) -> str:
+        """将source_code转换成template_func一样的入参和返回类型"""
+        prompt = CLONE_PROMPT.format(source_code=source_code, template_func=template_func)
+        logger.info(f"query for CloneFunction: \n {prompt}")
+        code = await self.write_code(prompt)
+        assert 'def' in code
```
Comment (stellaHSR) [inline metagpt/actions/clone_function.py:45]: It's not recommended to include assert statements in this function. Instead, you can add unit tests to test its functionality.

File: metagpt/tools/code_interpreter.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    return code
+
+
+class OpenCodeInterpreter(object):
+    """https://github.com/KillianLucas/open-interpreter"""
+    def __init__(self, auto_run: bool = True) -> None:
+        interpreter = Interpreter()
+        interpreter.auto_run = auto_run
+        interpreter.model = CONFIG.openai_api_model or "gpt-3.5-turbo"
+        interpreter.api_key = CONFIG.openai_api_key
```
Comment (stellaHSR) [inline metagpt/tools/code_interpreter.py:44]: Would it be better to initialize the interpreter instance directly using these parameters?

File: metagpt/tools/code_interpreter.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        return self.interpreter.chat(query, return_messages=True)
+
+    @staticmethod
+    def extract_function(query_respond: List, function_name: str, *, language: str = 'python',
+                         function_format: str = None) -> str:
+        """create a function from query_respond."""
+        if language not in ('python'):
+            raise NotImplementedError(f"Not support to parse language {language}!")
+
+        # 定义函数形式
```
Comment (stellaHSR) [inline metagpt/tools/code_interpreter.py:59]: also, provide an English version for the comments.

**Comment by stellaHSR** [comment]
(https://github.com/KillianLucas/open-interpreter/issues/72)  It is important to note that Open-interpreter is currently compatible with Python 3.9. Therefore, this feature or tool may require Python 3.10 to work properly.

File: metagpt/tools/code_interpreter.py (hunk 2 above)
Comment (orange-crow) [reply metagpt/tools/code_interpreter.py:43]: As of the current version 0.1.3, [Interpreter](https://github.com/KillianLucas/open-interpreter/blob/main/interpreter/interpreter.py)  only supports initialize the interpreter instance directly using these parameters. Here is the code for Interpreter
```python

class Interpreter:

  def __init__(self):
    self.messages = []
    self.temperature = 0.001
    self.api_key = None
    self.auto_run = False
    self.local = False
    self.model = "gpt-4"
    self.debug_mode = False
    ...
```

File: metagpt/actions/clone_function.py (hunk 1 above)
Comment (orange-crow) [reply metagpt/actions/clone_function.py]: Thank you for your feedback. I've removed the assert statements from the function as suggested. Instead, I've added unit tests to ensure the functionality of the code. This approach should help improve code maintainability and testing coverage. Please let me know if you have any further suggestions or if there's anything else you'd like me to address.

File: metagpt/tools/code_interpreter.py (hunk 3 above)
Comment (orange-crow) [reply metagpt/tools/code_interpreter.py]: Thank you for pointing that out. I've now provided English versions for all the comments in the code except the unit test. If you have any more feedback or if there's anything else that needs attention, please feel free to let me know.

File: tests/metagpt/actions/test_clone_function.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+template_code = """
+def stock_indicator(stock_path: str, indicators=['Simple Moving Average', 'BollingerBands', 'MACD]) -> pd.DataFrame:
+    import pandas as pd
+    # here is your code.
+"""
+
+
+def get_expected_res():
+    import pandas as pd
+    import ta
```
[the remark this answers is not in the document]
Comment (orange-crow) [reply tests/metagpt/actions/test_clone_function.py:31]: Added 'ta' to requirements.txt as requested.

File: tests/metagpt/actions/test_clone_function.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+@pytest.mark.asyncio
+async def test_clone_function():
+    clone = CloneFunction()
+    code = await clone.run(template_code, source_code)
+    assert 'def ' in code
+    stock_path = './tests/data/baba_stock.csv'
+    df, msg = run_fucntion_code(code, 'stock_indicator', stock_path)
+    assert not msg
+    expected_df = get_expected_res()
+    assert df.equals(expected_df)
```
[the remark this answers is not in the document]
Comment (orange-crow) [reply tests/metagpt/actions/test_clone_function.py:54]: It seems like the issue is due to the 'pandas' import statement being inside a function, causing 'pd' not to be defined when calling 'pd.DataFrame'. I have already made the necessary changes in the test code to address this issue.

**Comment by stellaHSR** [comment]
LGTM

**Comment by stellaHSR** [comment]
According to the official repository, Open-interpreter requires Python 3.10. So, if you intend to use OI, please ensure that you are using Python 3.10.

**Pull request merged**